gorapass/views.py

```diff
@@ -23,9 +35,49 @@ def hike(request, hike_id):
 
 def hikes(request):
     hike_models = Hikes.objects.all()
+
+    # Filter out hikes if there is filtration criteria in the request
+    if request.body:
+        filters = json.loads(request.body)['filters']
+        if not valid_hike_filters(filters):
+            return HttpResponseBadRequest('Invalid filtration criteria')
+
+        hike_models = filter_hikes(hike_models, filters)
+
     hike_dicts = [ model_to_dict(hike) for hike in hike_models ]
     return JsonResponse(hike_dicts, safe=False)
 
+def valid_hike_filters(filters):
+    for filter in filters:
+        if set(filter.keys()) != FILTER_KEYS:
+            return False
+        if not hasattr(Hikes, filter[ATTRIBUTE_NAME]):
+            return False
+        
+    return True
+
+def filter_hikes(hike_models, filters):
+    for filter in filters:
+        attribute = filter[ATTRIBUTE_NAME]
+        value = filter[ATTRIBUTE_VALUE]
+        match filter[FILTER_TYPE]:
+            case 'exact':
+                hike_models = [ hike for hike in hike_models if getattr(hike, attribute) == value ]
+            case 'partial':
+                hike_models = [ hike for hike in hike_models if value.lower() in str(getattr(hike, attribute)).lower() ]
+            case 'less_than':
+                hike_models = [ hike for hike in hike_models if getattr(hike, attribute) < value ]
+            case 'greater_than':
+                hike_models = [ hike for hike in hike_models if getattr(hike, attribute) > value ]
+                pass
+            case _:
```

Comment:
What is `case _:` used for?

Reply:
Ahh, I forgot about that. In theory, we shouldn't make it to the default case, but if we do we should probably return a 400 bad request or a 500 server error. The validation function should catch any filters that aren't present. 

I think for now we can roll with this so we don't block things, but let's add something like "validate filter types" as a todo.